flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1058,6 +1092,30 @@ public void handleError(Throwable throwable) {
 		}
 	}
 
+	/**
+	 * The heartbeat listener for JobManager and ResourceManager, they can be distinguished by ResourceID
+	 * and trigger different processes.
+	 */
+	private final class JMRMHeartbeatListener implements HeartbeatListener {
```

Comment:
Hmm it could make sense to have a version specialized for the JM and RM because you will trigger different operations. And then the TM doesn't have to make the distinction.